tests/test_username_api.py

```diff
@@ -40,405 +45,87 @@ def get_expected_response(website, user, status):
-      expected = get_expected_response('twitter', user, 404)
-      actual = get_response(self.app, 'twitter', user)
-
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
-      else:
-        message += ' but {} is still available'.format(user)
-
-    if message is not None:
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['facebook']['available_usernames'])
-  def test_available_facebook_username(self, user):
-    expected = get_expected_response('facebook', user, 404)
-    actual = get_response(self.app, 'facebook', user)
+  @parameterized.expand(load_availability_test_cases('available_username'),
+      testcase_func_name=custom_name_func)
+  def test_available_username(self, website, user):
+    expected = get_expected_response(website, user, 404)
+    actual = get_response(self.app, website, user)
 
     # this is special to facebook checking,
     # and 'profile' field is only useful for
     # frontend purposes. Skipping here.
-    del actual['profile']
-
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(25)
-      expected = get_expected_response('facebook', user, 404)
-      actual = get_response(self.app, 'facebook', user)
-
+    if website == 'facebook':
       del actual['profile']
 
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
-      else:
-        message += ' but {} is still available'.format(user)
-
-    if message is not None:
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['openhub']['available_usernames'])
-  def test_available_openhub_username(self, user):
-    expected = get_expected_response('openhub', user, 404)
-    actual = get_response(self.app, 'openhub', user)
-
     message = None
     if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(30, first_char_letter=True)
-      expected = get_expected_response('openhub', user, 404)
-      actual = get_response(self.app, 'openhub', user)
-
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
-      else:
-        message += ' but {} is still available'.format(user)
-
-    if message is not None:
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['deviantart']['available_usernames'])
-  def test_available_deviantart_username(self, user):
-    expected = get_expected_response('deviantart', user, 404)
-    actual = get_response(self.app, 'deviantart', user)
+      username_pattern = self.websites['username_patterns'][website]
+      random_length = random.randint(username_pattern['min_length'],
```

Comment:
improving the generation of usernames is not part of this issue.
please remove it.
focus your commits.
We can create separate tasks for things like this.

Reply:
@jayvdb can you create a separate task?

Reply:
then I'll set length that just fit with every website.